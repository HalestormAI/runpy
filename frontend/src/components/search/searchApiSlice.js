import {createSlice} from "@reduxjs/toolkit";
import {call} from "../../utils/api";


// FIXME: For now, I'm hacking this in. The search API is due to change
// (e.g. https://trello.com/c/UZ5za0nt/1-graphql-for-better-querying)
// so this should be a stop-gap once I go back to the back-end
function buildSearchUrl(state) {
    const distanceForm = values => {
        return `/search/distance/${values.low}/${values.high}`
    };

    let builder = null;
    if (state.searchForm.type === "distance") {
        builder = distanceForm;
    } else {
        throw Error("Invalid URL builder type for search");
    }

    return builder(state.searchForm.values);
}

export const slice = createSlice({
    name: 'searchData',
    initialState: {
        activities: [],
        data: {
            errorMessage: null,
            waiting: false
        }
    },
    reducers: {
        searchFetch: state => {
            state.data.waiting = true;
        },
        searchDone: state => {
            state.data.waiting = false;
        },
        searchSuccess: (state, action) => {
            // TODO: Parse out the JSON and store into the state
            console.log(action.payload);
            state.data.errorMessage = null;
        },
        searchError: (state, action) => {
            state.data.waiting = false;
            state.data.errorMessage = action.payload.message;
        },
        searchClearError: state => {
            state.data.errorMessage = null;
        }
    }
});


export const {searchFetch, searchDone, searchSuccess, searchError, searchClearError} = slice.actions;

export const performSearch = () => (dispatch, getState) => {
    // const url = `/search/distance/${form_elem.distance_low.value}/${form_elem.distance_high.value}`;
    const url = buildSearchUrl(getState());
    dispatch(searchFetch());
    call(dispatch, url, searchDone, searchSuccess, searchError)
};

export default slice.reducer;
export const selectApiState = state => state.searchApi.data;