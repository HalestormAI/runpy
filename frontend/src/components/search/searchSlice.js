import {createSlice} from "@reduxjs/toolkit";
import {call} from "../../utils/api";


export const slice = createSlice({
    name: 'searchData',
    initialState: {
        activities: [],
        errorMessage: null,
        waiting: false
    },
    reducers: {
        searchFetch: state => {
            state.waiting = true;
        },
        searchDone: state => {
            state.waiting = false;
        },
        searchSuccess: (state, action) => {
            // TODO: Parse out the JSON and store into the state
            console.log(action.payload);
            state.errorMessage = null;
        },
        searchError: (state, action) => {
            state.waiting = false;
            state.errorMessage = action.payload.message;
        },
        searchClearError: state => {
            state.errorMessage = null;
        }
    }
});


export const {searchFetch, searchDone, searchSuccess, searchError, searchClearError} = slice.actions;

export const search = () => dispatch => {
    // const url = `/search/distance/${form_elem.distance_low.value}/${form_elem.distance_high.value}`;

    dispatch(searchFetch());
    call(dispatch, apiUri, searchDone, searchSuccess, searchError)
};

export default slice.reducer;
export const selectSearchWaiting = state => state.search.waiting;
export const selectSearchErrorMessage = state => state.search.errorMessage;
export const selectSearchActivities = state => state.search.activities;