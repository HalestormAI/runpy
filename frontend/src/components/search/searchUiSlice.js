import {createSlice} from "@reduxjs/toolkit";


export const slice = createSlice({
    name: 'searchData',
    initialState: {
        type: "distance",
        values: {
            "low": "",
            "high": ""
        }
    },
    reducers: {
        updateLow: (state, action) => {
            state.values.low = action.payload
        },
        updateHigh: (state, action) => {
            state.values.high = action.payload
        },
        setType: (state, action) => {
            if (action.payload !== "distance") {
                throw Error("Only distance search is implemented so far..")
            }
        }
    }
});


export const {updateLow, updateHigh, setType} = slice.actions;

export default slice.reducer;

export const selectFormState = state => state.searchForm;
