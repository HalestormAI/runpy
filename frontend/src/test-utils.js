// https://gist.github.com/6174/6062387
export const randomString = () => Math.random()
    .toString(36)
    .substring(2, 15) + Math.random()
    .toString(36)
    .substring(2, 15);


// https://gist.github.com/miguelmota/5b67e03845d840c949c4
export const randomDate = (start, end) => {
    return new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()));
}

export const createMockActivity = (overrides) => ({
    name: randomString(),
    // Make sure we always generate a value > 0
    average_speed: 3 * Math.random() + 2,
    distance: 42.2 * Math.random() + Number.MIN_VALUE,
    workout_type: "Run",
    start_date: randomDate(new Date(2012, 0, 1), new Date()),
    total_elevation_gain: 250 * Math.random(),
    moving_time: 3000 * Math.random() + 600,
    ...overrides
});

export const createMockActivityArray = (num) => ([...Array(num).keys()]
    .reduce((acc, x) => {
        acc.push(createMockActivity());
        return acc;
    }, []));