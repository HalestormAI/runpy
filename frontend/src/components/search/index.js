import React from "react";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import SearchIcon from '@material-ui/icons/Search';
import {useDispatch, useSelector} from "react-redux";
import {isValidFormType, selectFormState} from "./searchUiSlice";
import {performSearch, selectApiState} from "./searchApiSlice";
import DistanceSearchFields from "./filterTypes/searchTypeDistance";
import CircularProgress from "@material-ui/core/CircularProgress";
import makeStyles from "@material-ui/core/styles/makeStyles";
import {grey} from '@material-ui/core/colors';

const useStyles = makeStyles(theme => ({
    root: {
        display: 'flex',
        alignItems: 'center',
    },
    wrapper: {
        // margin: theme.spacing(1),
        position: 'relative',
    },
    buttonProgress: {
        color: grey[500],
        position: 'absolute',
        top: '50%',
        left: '50%',
        marginTop: -12,
        marginLeft: -12,
    }
}));

export function SearchFormComponent(props) {
    const {formType, registerValidationFields, updateValidationState} = props;
    return (<React.Fragment>
        {formType === "distance" &&
        <DistanceSearchFields registerValidationFields={registerValidationFields}
                              updateValidationState={updateValidationState}/>
        }
    </React.Fragment>)
}

export const formIsValid = state => Object.values(state)
    .reduce((isValid, fieldValid) => isValid && fieldValid, true);

export default function StravaSearchComponent() {
    const dispatch = useDispatch();
    const formState = useSelector(selectFormState);
    const apiState = useSelector(selectApiState);

    const [validState, setValidState] = React.useState({});

    const handleSubmit = e => {
        e.preventDefault();
        if (formIsValid(validState)) {
            dispatch(performSearch());
        }

        return false;
    };

    // TODO: This is pretty hairy, probably needs rethinking...
    const registerValidationFields = (names) => {
        setValidState(names.reduce((nameValid, name) => ({
            ...nameValid,
            [name]: false
        }), validState));
    };

    const updateFieldValidation = (name, isValid) => {
        setValidState({...validState, [name]: isValid});
    };

    const classes = useStyles();

    return (
        <form onSubmit={handleSubmit} autoComplete="off">
            <Grid container spacing={3}>
                <Grid item xs={1}>
                    Search:
                </Grid>

                {isValidFormType(formState.type) &&
                <SearchFormComponent formType={formState.type}
                                     registerValidationFields={registerValidationFields}
                                     updateValidationState={updateFieldValidation}/>
                }
                <Grid item xs={1} className={classes.root}>
                    <div className={classes.wrapper}>
                        <Button
                            variant="contained"
                            color="primary"
                            type="submit"
                            endIcon={<SearchIcon/>}
                            disabled={apiState.waiting}
                            className="search-button"
                        >
                            Search
                        </Button>
                        {apiState.waiting &&
                        <CircularProgress size={24} className={classes.buttonProgress} color="primary"/>}
                    </div>
                </Grid>
            </Grid>
        </form>
    )
}
