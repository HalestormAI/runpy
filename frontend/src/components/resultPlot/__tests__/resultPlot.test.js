import React from "react";
import Plot from '../../../../node_modules/react-plotly.js/react-plotly';
import {shallow} from 'enzyme'
import SearchResultPlotComponent, {updatePlotState} from "../index";
import {useSelector} from "react-redux";
import {initialState as uiInitialState, selectFormState} from "../../search/searchUiSlice";
import {selectActivities} from "../../search/searchApiSlice";
import {randomString} from "../../../test-utils";

jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
}));

describe("the result plot component", () => {

    describe("no activities", () => {

        let mockAppStates = null;

        beforeEach(() => {
            useSelector.mockImplementation(callback => {
                return mockAppStates[callback];
            });
            mockAppStates = {
                [selectFormState]: {
                    ...uiInitialState
                },
                [selectActivities]: []
            }
        });

        afterEach(() => {
            jest.resetAllMocks();
        });

        afterEach(() => {
            jest.resetAllMocks();
        })

        it('should render', () => {
            const wrapper = shallow(<SearchResultPlotComponent/>);
            const plot = wrapper.find(Plot);
            expect(plot).toMatchSnapshot();
            const props = plot.props();
            expect(props.data[0].x).toHaveLength(0);
            expect(props.data[0].y).toHaveLength(0);
            expect(props.data[0].text).toHaveLength(0);
            expect(props.layout.title).toEqual("Pace (mins/km) over time 0.00km ... 0.00km");
            expect(props.layout.yaxis.tickformat).toEqual("%M:%S");
            expect(props.layout.xaxis.tickformat).toEqual("%Y-%m-%d");
        });
    });

    describe('plot title', () => {

        let mockAppStates = null;

        beforeEach(() => {
            useSelector.mockImplementation(callback => {
                return mockAppStates[callback];
            });
            mockAppStates = {
                [selectFormState]: {
                    ...uiInitialState,
                    values: {
                        low: 1234,
                        high: 5678
                    }
                },
                [selectActivities]: []
            }
        });


        it('should have a title that reflects the low/high bounds', () => {
            const wrapper = shallow(<SearchResultPlotComponent/>);
            const plot = wrapper.find(Plot);
            const props = plot.props();
            expect(props.layout.title).toEqual("Pace (mins/km) over time 1.23km ... 5.68km");
        })
    });

    describe("some activities", () => {

        const mockAppStates = {
            [selectFormState]: {
                ...uiInitialState
            },
            [selectActivities]: [1, 2, 3].map(() => ({
                average_speed: 5 * Math.random(),
                workout_type: randomString(),
                start_date: Date.now().toLocaleString()
            }))
        }

        beforeEach(() => {
            useSelector.mockImplementation(callback => {
                return mockAppStates[callback];
            });
        });

        afterEach(() => {
            jest.resetAllMocks();
        })

        afterEach(() => {
            jest.resetAllMocks();
        });

        it('should render', () => {
            const wrapper = shallow(<SearchResultPlotComponent/>);
            const plot = wrapper.find(Plot);
            const props = plot.props();
            expect(props.data[0].x).toHaveLength(3);
            expect(props.data[0].y).toHaveLength(3);
            expect(props.data[0].text).toHaveLength(3);
            expect(props.layout.title).toEqual("Pace (mins/km) over time 0.00km ... 0.00km");
            expect(props.layout.yaxis.tickformat).toEqual("%M:%S");
            expect(props.layout.xaxis.tickformat).toEqual("%Y-%m-%d");
        });

        it('should update the state when update fires', () => {
            const mockTitle = randomString();

            const mockState = {
                layout: {
                    initial: "layoutValue",
                    overwriteme: "original"
                },
                config: {},
                frames: {}
            }

            const mockLayoutMemo = {
                yaxis: {
                    tickformat: "%M:%S"
                },
                xaxis: {
                    tickformat: "%Y-%m-%d"
                }
            };

            const mockFig = {
                layout: {
                    title: mockTitle,
                    overwriteme: "overwriteme"
                },
                config: {
                    configKey: "configValue"
                },
                frames: {
                    framesKey: "framesValue"
                }
            };

            expect(mockState.layout.initial).toEqual("layoutValue");
            expect(mockState.layout.overwriteme).toEqual("original");
            expect(mockState.layout).not.toHaveProperty("title");
            expect(mockState.layout).not.toHaveProperty("yaxis");
            expect(mockState.layout).not.toHaveProperty("xaxis");
            expect(mockState.config).toEqual({});
            expect(mockState.frames).toEqual({});

            const setState = jest.fn(newState => {
                console.log(newState)
                expect(newState.layout.initial).toEqual("layoutValue");
                expect(newState.layout.overwriteme).toEqual("overwriteme");
                expect(newState.layout.title).toEqual(mockTitle);
                expect(newState.layout.yaxis.tickformat).toEqual(mockLayoutMemo.yaxis.tickformat);
                expect(newState.layout.xaxis.tickformat).toEqual(mockLayoutMemo.xaxis.tickformat);
                expect(newState.config.configKey).toEqual("configValue");
                expect(newState.frames.framesKey).toEqual("framesValue");
            })

            updatePlotState(mockState, setState, mockFig, mockLayoutMemo);
        });
    });

});