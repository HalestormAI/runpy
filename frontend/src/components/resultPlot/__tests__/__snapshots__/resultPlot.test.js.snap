// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the result plot component no activities should render 1`] = `
<PlotlyComponent
  data={
    Array [
      Object {
        "mode": "markers",
        "text": Array [],
        "transforms": Array [
          Object {
            "groups": Array [],
            "type": "groupby",
          },
        ],
        "type": "scatter",
        "x": Array [],
        "y": Array [],
      },
    ]
  }
  debug={false}
  layout={
    Object {
      "autosize": true,
      "font": Object {
        "color": "rgba(0, 0, 0, 0.87)",
      },
      "paper_bgcolor": "#fff",
      "plot_bgcolor": "#fff",
      "responsize": true,
      "title": "Pace (mins/km) over time 0.00km ... 0.00km",
      "xaxis": Object {
        "gridcolor": "#fafafa",
        "linecolor": "#fafafa",
        "tickformat": "%Y-%m-%d",
        "zerolinecolor": "#fafafa",
      },
      "yaxis": Object {
        "gridcolor": "#fafafa",
        "linecolor": "#fafafa",
        "tickformat": "%M:%S",
        "zerolinecolor": "#fafafa",
      },
    }
  }
  onInitialized={[Function]}
  onUpdate={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  useResizeHandler={true}
/>
`;
